packages/firestore/src/local/query_engine.ts

```diff
@@ -190,66 +215,135 @@ export class QueryEngine {
+      return PersistencePromise.resolve<DocumentMap | null>(null);
     }
 
-    if (getLogLevel() <= LogLevel.DEBUG) {
-      logDebug(
-        'QueryEngine',
-        'Query:',
-        stringifyQuery(query),
-        'scans',
-        context.documentReadCount,
-        'local documents and returns',
-        resultSize,
-        'documents as results.'
-      );
+    // Queries that have never seen a snapshot without limbo free documents
+    // should also be run as a full collection scan.
+    if (lastLimboFreeSnapshotVersion.isEqual(SnapshotVersion.min())) {
+      return PersistencePromise.resolve<DocumentMap | null>(null);
     }
 
-    if (
-      context.documentReadCount >
-      this.relativeIndexReadCostPerDocument * resultSize
-    ) {
-      if (getLogLevel() <= LogLevel.DEBUG) {
-        logDebug(
-          'QueryEngine',
-          'The SDK decides to create cache indexes for query:',
-          stringifyQuery(query),
-          'as using cache indexes may help improve performance.'
-        );
+    return this.localDocumentsView!.getDocuments(transaction, remoteKeys).next(
+      documents => {
+        const previousResults = applyQuery(query, documents);
+
+        if (
+          needsRefill(
+            query,
+            previousResults,
+            remoteKeys,
+            lastLimboFreeSnapshotVersion
+          )
+        ) {
+          return PersistencePromise.resolve<DocumentMap | null>(null);
+        }
+
+        if (getLogLevel() <= LogLevel.DEBUG) {
+          logDebug(
+            'QueryEngine',
+            'Re-using previous result from %s to execute query: %s',
+            lastLimboFreeSnapshotVersion.toString(),
+            stringifyQuery(query)
+          );
+        }
+
+        // Retrieve all results for documents that were updated since the last
+        // limbo-document free remote snapshot.
+        return appendRemainingResults(
+          transaction,
+          this.localDocumentsView,
+          previousResults,
+          query,
+          newIndexOffsetSuccessorFromReadTime(
+            lastLimboFreeSnapshotVersion,
+            INITIAL_LARGEST_BATCH_ID
+          )
+        ).next<DocumentMap | null>(results => results);
       }
-      return this.indexManager.createTargetIndexes(
-        transaction,
-        queryToTarget(query)
+    );
+  }
+
+  private executeFullCollectionScan(
+    transaction: PersistenceTransaction,
+    query: Query,
+    context: QueryContext
+  ): PersistencePromise<DocumentMap> {
+    if (getLogLevel() <= LogLevel.DEBUG) {
+      logDebug(
+        'QueryEngine',
+        'Using full collection scan to execute query:',
+        stringifyQuery(query)
       );
     }
 
-    return PersistencePromise.resolve();
+    return this.localDocumentsView!.getDocumentsMatchingQuery(
+      transaction,
+      query,
+      IndexOffset.min(),
+      context
+    );
   }
+}
+
+export class QueryEngineFieldIndexPlugin {
```

Comment:
There is another "plugin" class in the next PR, currently named `IndexManagerFieldIndexPlugin`. IMO it would improve readability if their names followed the same pattern, and `IndexedIndexManagerPlugin` seems a bit weird. What do you think? What would be your ideal names for these classes given that they both are "plugins" for CSI?

Reply:
See the changes to `index_manager.ts` in https://github.com/firebase/firebase-js-sdk/pull/7951 for the exact code to which I'm referring.

Reply:
OK, fair.